### PR TITLE
Change: switch to internal fandom wiki search engine, increase scope of skyhanni's /wiki command to include any item whose item lore includes wiki.

### PR DESCRIPTION
**Pull request by RayDeeUx**
overrides #485 from my alt account @ErymanthusRayDeeUx.

### CODE REVIEW

**Comment by hannibal002** [comment]
Does this mean https://github.com/hannibal002/SkyHanni/pull/485 can be closed?

**Comment by RayDeeUx** [comment]
> Does this mean #485 can be closed?

eh sure why not @hannibal002

File: src/main/java/at/hannibal2/skyhanni/features/misc/FandomWikiIsSoMuchBetter.kt [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import io.github.moulberry.notenoughupdates.events.SlotClickEvent
+import net.minecraft.client.Minecraft
+import net.minecraft.client.gui.inventory.GuiChest
+import net.minecraft.inventory.ContainerChest
+import net.minecraft.inventory.Slot
+import net.minecraft.item.ItemStack
+import net.minecraftforge.fml.common.eventhandler.EventPriority
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
+
+class FandomWikiIsSoMuchBetter {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/misc/FandomWikiIsSoMuchBetter.kt]: Please consider using a more descriptive class name instead.

File: src/main/java/at/hannibal2/skyhanni/features/misc/FandomWikiIsSoMuchBetter.kt [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    fun onSlotClick(event: SlotClickEvent) {
+
+        if (!inInventory || !isEnabled()) return
+
+        val chestName = InventoryUtils.openInventoryName()
+        val itemName = event.slot.stack.displayName
+        val itemInHand = InventoryUtils.getItemInHand()?.name ?: ""
+
+        if ((itemInHand == "") || (event.slotId == 11 && itemName.contains("Wiki Command") && chestName.contains("Wiki"))) {
+            LorenzUtils.clickableChat("§e[SkyHanni] Click here to visit the Hypixel Skyblock Fandom Wiki!", "shwiki")
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/misc/FandomWikiIsSoMuchBetter.kt]: Why do we need this /shwiki command in the first place? cant we use /wiki here ?

File: src/main/java/at/hannibal2/skyhanni/features/misc/FandomWikiIsSoMuchBetter.kt [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    @SubscribeEvent(priority = EventPriority.HIGH)
+    fun onSlotClick(event: SlotClickEvent) {
+
+        if (!inInventory || !isEnabled()) return
+
+        val chestName = InventoryUtils.openInventoryName()
+        val itemName = event.slot.stack.displayName
+        val itemInHand = InventoryUtils.getItemInHand()?.name ?: ""
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/misc/FandomWikiIsSoMuchBetter.kt]: Maybe use .nameWithEnchantment instead of .name here

**Review by hannibal002 — COMMENTED** [review]
Also idk if It's still in the scope of this pr, but it would be cool if we have a hotkey to search for an item in the wiki by hovering over the item and pressing the hotkey (think of those debug keybinds in the screenshot) 
image

**Comment by hannibal002** [comment]
Marking as draft until the suggested changes are addressed.

File: src/main/java/at/hannibal2/skyhanni/features/misc/FandomWikiIsSoMuchBetter.kt (hunk 3 above)
Comment (RayDeeUx) [reply src/main/java/at/hannibal2/skyhanni/features/misc/FandomWikiIsSoMuchBetter.kt]: done

File: src/main/java/at/hannibal2/skyhanni/config/features/CommandsConfig.java [hunk 4]
```diff
@@ -56,11 +58,24 @@ public static class TabCompleteConfig {
...
-    public boolean useFandomWiki = false;
+    public FandomWikiCommmandConfig fandomWiki = new FandomWikiCommmandConfig();
+
+    public static class FandomWikiCommmandConfig {
+
+        @Expose
+        @ConfigOption(name = "Use Fandom", desc = "Use Fandom wiki (§ehypixel-skyblock.fandom.com§7) instead of the Hypixel wiki (§ewiki.hypixel.net§7).")
+        @ConfigEditorBoolean
+        @FeatureToggle
+        public boolean useFandomWiki = false;
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/config/features/CommandsConfig.java:84]: Please add config migration for this. also maybe replace the name to something better, like "replace chat" and use this as the name of this toggle as well

File: src/main/java/at/hannibal2/skyhanni/features/misc/FandomWikiFromMenus.kt [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+import net.minecraftforge.fml.common.eventhandler.EventPriority
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
+
+class FandomWikiFromMenus {
+
+    @SubscribeEvent(priority = EventPriority.HIGH)
+    fun onSlotClick(event: SlotClickEvent) {
+        val chestName = InventoryUtils.openInventoryName()
+
+        if (chestName.isEmpty() || !isEnabled() || !LorenzUtils.inSkyBlock) return
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/misc/FandomWikiFromMenus.kt]: Please check if on skyblock before check if the feature is enabled

File: src/main/java/at/hannibal2/skyhanni/features/misc/FandomWikiFromMenus.kt [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import net.minecraft.inventory.Slot
+import net.minecraft.item.ItemStack
+import net.minecraftforge.fml.common.eventhandler.EventPriority
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
+
+class FandomWikiFromMenus {
+
+    @SubscribeEvent(priority = EventPriority.HIGH)
+    fun onSlotClick(event: SlotClickEvent) {
+        val chestName = InventoryUtils.openInventoryName()
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/misc/FandomWikiFromMenus.kt:33]: please only use this "heavy" function after confirming this feature is enabled

File: src/main/java/at/hannibal2/skyhanni/features/misc/FandomWikiFromMenus.kt [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        if (chestName.isEmpty() || !isEnabled() || !LorenzUtils.inSkyBlock) return
+
+        val itemClickedStack = event.slot.stack
+        val itemClickedName = itemClickedStack.displayName
+        val itemInHand = InventoryUtils.getItemInHand() ?: return
+        val itemInHandName = itemInHand.nameWithEnchantment ?: return
+        val internalName = itemInHand.getInternalName().asString() ?: return
+
+        var placeHolderOne = ""
+        var placeHolderTwo = ""
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/misc/FandomWikiFromMenus.kt]: Please use better names: wiki internal name and display name?

File: src/main/java/at/hannibal2/skyhanni/features/commands/WikiCommand.kt [hunk 8]
```diff
@@ -1,34 +1,75 @@
...
+import at.hannibal2.skyhanni.utils.NEUItems
 import at.hannibal2.skyhanni.utils.OSUtils
+import at.hannibal2.skyhanni.utils.StringUtils.removeColor
 import net.minecraft.network.play.client.C01PacketChatMessage
+import net.minecraft.client.Minecraft
+import net.minecraft.item.ItemStack
 import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
+import org.lwjgl.input.Keyboard
 
 class WikiCommand {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/commands/WikiCommand.kt]: Maybe rename the whole class to WikiReplacer or Wiki Manager or something, its more than a command at this point

File: src/main/java/at/hannibal2/skyhanni/features/misc/FandomWikiFromMenus.kt [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+            LorenzUtils.clickableChat("§e[SkyHanni] Click here to visit the Hypixel Skyblock Fandom Wiki!", "wiki")
+            return
+        } else if (event.slotId == 15 && itemClickedName.contains("Wikithis Command") && chestName.contains("Wiki")) {
+            placeHolderOne = itemInHandName
+            placeHolderTwo = internalName
+        } else if ((itemClickedStack.getLore().anyContains("Wiki") || itemClickedStack.getLore().anyContains("wiki")) && !(itemClickedStack.getLore().anyContains("wikipedia"))) { //.lowercase() to match "Wiki!" and ".*wiki.*" lore lines in one fell swoop
+            placeHolderOne = itemClickedName.removeColor().replace("✔ ", "").replace("✖ ", "")
+            placeHolderTwo = itemClickedName.removeColor().replace("✔ ", "").replace("✖ ", "")
+        } else return
+        LorenzUtils.clickableChat("§e[SkyHanni] Click here to search for $placeHolderOne §eon the Hypixel Skyblock Fandom Wiki!", "wiki $placeHolderTwo")
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/misc/FandomWikiFromMenus.kt]: Can we maybe just ignore the chat message entirely and just directy opening the wiki at this point? the original hypixel behaviour always felt clucky to me.

**Review by hannibal002 — CHANGES_REQUESTED** [review]
for examples on how to properly migrate the config, see the revsion/commit fe505fc83c2e608e6c30e36568bb6a1f2ecfade0

File: src/main/java/at/hannibal2/skyhanni/features/misc/FandomWikiFromMenus.kt (hunk 7 above)
Comment (RayDeeUx) [reply src/main/java/at/hannibal2/skyhanni/features/misc/FandomWikiFromMenus.kt]: done

File: src/main/java/at/hannibal2/skyhanni/features/commands/WikiCommand.kt (hunk 8 above)
Comment (RayDeeUx) [reply src/main/java/at/hannibal2/skyhanni/features/commands/WikiCommand.kt]: done

File: src/main/java/at/hannibal2/skyhanni/features/misc/FandomWikiFromMenus.kt (hunk 6 above)
Comment (RayDeeUx) [reply src/main/java/at/hannibal2/skyhanni/features/misc/FandomWikiFromMenus.kt:35]: done

File: src/main/java/at/hannibal2/skyhanni/features/misc/FandomWikiFromMenus.kt (hunk 5 above)
Comment (RayDeeUx) [reply src/main/java/at/hannibal2/skyhanni/features/misc/FandomWikiFromMenus.kt]: done

File: src/main/java/at/hannibal2/skyhanni/features/misc/FandomWikiFromMenus.kt (hunk 9 above)
Comment (RayDeeUx) [reply src/main/java/at/hannibal2/skyhanni/features/misc/FandomWikiFromMenus.kt]: done as a separate config option toggle

File: src/main/java/at/hannibal2/skyhanni/config/features/CommandsConfig.java (hunk 4 above)
Comment (RayDeeUx) [reply src/main/java/at/hannibal2/skyhanni/config/features/CommandsConfig.java:78]: done

**Comment by RayDeeUx** [comment]
~~no walker im not done yet~~

**Review by hannibal002 — APPROVED** [review]

**Pull request merged**